### PR TITLE
Draft of support for vectors in binary and ternary expressions

### PR DESCRIPTION
**Pull request by Jane-Of-Art**
Draft of the changes I'm proposing for supporting vectors in all binary expressions (min, max, pow, atan2). Unfortunately I couldn't get my local machine to run and check everything yet so I will have to see if all new specs pass!

This is a part of ticket #4865

@lilleyse @Dylan-Brown

### CODE REVIEW

**Comment by Jane-Of-Art** [comment]
I'm not sure if it's my throw new DeveloperError notation that's wrong, but I don't see why all of my Error-throwing tests are failing :(

The good news is, properly-formated expression calls seem to all be passing!

```
  Scene/Expression
    ✗ throws if atan2 function takes invalid types / instances of arguments
	Expected function to throw DeveloperError.
	    at Specs/Scene/ExpressionSpec.js:2110:12
	    at Object.<anonymous> (Specs/customizeJasmine.js:80:30)

	Expected function to throw DeveloperError.
	    at Specs/Scene/ExpressionSpec.js:2114:12
	    at Object.<anonymous> (Specs/customizeJasmine.js:80:30)

	Expected function to throw DeveloperError.
	    at Specs/Scene/ExpressionSpec.js:2118:12
	    at Object.<anonymous> (Specs/customizeJasmine.js:80:30)

    ✗ throws if pow function takes invalid types / instances of arguments
	Expected function to throw DeveloperError.
	    at Specs/Scene/ExpressionSpec.js:2152:12
	    at Object.<anonymous> (Specs/customizeJasmine.js:80:30)

	Expected function to throw DeveloperError.
	    at Specs/Scene/ExpressionSpec.js:2156:12
	    at Object.<anonymous> (Specs/customizeJasmine.js:80:30)

	Expected function to throw DeveloperError.
	    at Specs/Scene/ExpressionSpec.js:2160:12
	    at Object.<anonymous> (Specs/customizeJasmine.js:80:30)

    ✗ throws if min function takes invalid types / instances of arguments
	Expected function to throw DeveloperError.
	    at Specs/Scene/ExpressionSpec.js:2196:12
	    at Object.<anonymous> (Specs/customizeJasmine.js:80:30)

	Expected function to throw DeveloperError.
	    at Specs/Scene/ExpressionSpec.js:2200:12
	    at Object.<anonymous> (Specs/customizeJasmine.js:80:30)

    ✗ throws if min function takes invalid types / instances of arguments
	Expected function to throw DeveloperError.
	    at Specs/Scene/ExpressionSpec.js:2236:12
	    at Object.<anonymous> (Specs/customizeJasmine.js:80:30)

	Expected function to throw DeveloperError.
	    at Specs/Scene/ExpressionSpec.js:2240:12
	    at Object.<anonymous> (Specs/customizeJasmine.js:80:30)
```

**Comment by Jane-Of-Art** [comment]
@lilleyse 

Dylan just took a look and couldn't figure out why these cases aren't throwing errors properly. Could you take a look at my syntax quickly?

Thanks!

**Comment by lilleyse** [comment]
Sure, I'll take a look at this later today.

File: Specs/Scene/ExpressionSpec.js [hunk 1]
```diff
@@ -2077,6 +2077,21 @@ defineSuite([
...
+        expression = new Expression('atan2(vec3(0,1,0.5),vec3(1,0,0.5))');
+        expect(expression.evaluate(frameState, undefined).x).toEqualEpsilon(0.0, CesiumMath.EPSILON10);
+        expect(expression.evaluate(frameState, undefined).y).toEqualEpsilon(0.5 * Math.PI, CesiumMath.EPSILON10);
+        expect(expression.evaluate(frameState, undefined).z).toEqualEpsilon(0.25 * Math.PI, CesiumMath.EPSILON10);
+
+        expression = new Expression('atan2(vec4(0,1,0.5,1),vec4(1,0,0.5,0))');
+        expect(expression.evaluate(frameState, undefined).x).toEqualEpsilon(0.0, CesiumMath.EPSILON10);
+        expect(expression.evaluate(frameState, undefined).y).toEqualEpsilon(0.5 * Math.PI, CesiumMath.EPSILON10);
+        expect(expression.evaluate(frameState, undefined).z).toEqualEpsilon(0.25 * Math.PI, CesiumMath.EPSILON10);
+        expect(expression.evaluate(frameState, undefined).w).toEqualEpsilon(0.5 * Math.PI, CesiumMath.EPSILON10);
```
Comment (lilleyse) [inline Specs/Scene/ExpressionSpec.js]: For each of these, just evaluate the expression once and do a `toEqualEpsilon` on a Cartesian, like `evaluates pow function` below.

File: Specs/Scene/ExpressionSpec.js [hunk 2]
```diff
@@ -2089,12 +2104,36 @@ defineSuite([
...
+
+        expression = new Expression('pow(vec2(5,4),vec2(0,2))');
+        expect(expression.evaluate(frameState, undefined)).toEqual(new Cartesian2(1.0, 16.0));
+
+        expression = new Expression('pow(vec3(5,4,3),vec3(0,2,3))');
+        expect(expression.evaluate(frameState, undefined)).toEqual(new Cartesian3(1.0, 16.0, 27.0));
+
+        expression = new Expression('pow(vec4(5,4,3,2),vec4(0,2,3,5))');
+        expect(expression.evaluate(frameState, undefined)).toEqual(new Cartesian4(1.0, 16.0, 27.0, 32.0));
+
```
Comment (lilleyse) [inline Specs/Scene/ExpressionSpec.js]: Remove empty line.

File: Specs/Scene/ExpressionSpec.js [hunk 3]
```diff
@@ -2089,12 +2104,36 @@ defineSuite([
         }).toThrowDeveloperError();
     });
 
+    it('throws if atan2 function takes invalid types / instances of arguments', function() {
```
Comment (lilleyse) [inline Specs/Scene/ExpressionSpec.js]: Just `throws if atan2 function takes mismatching types` is fine.

Comment (lilleyse) [inline Specs/Scene/ExpressionSpec.js]: Same comment for the other tests.

File: Specs/Scene/ExpressionSpec.js [hunk 4]
```diff
@@ -2143,6 +2230,17 @@ defineSuite([
...
+        expect(function() {
+            return new Expression('max(0.0, vec2(0,1))');
+        }).toThrowDeveloperError();
+
+        expect(function() {
+            return new Expression('max(vec2(0,1),vec3(0,1,2))');
+        }).toThrowDeveloperError();
+    });
+
+
```
Comment (lilleyse) [inline Specs/Scene/ExpressionSpec.js]: Remove empty line.

File: Source/Scene/Expression.js [hunk 5]
```diff
@@ -79,10 +79,10 @@ define([
...
 
     var binaryFunctions = {
-        atan2 : Math.atan2,
-        pow : Math.pow,
-        min : Math.min,
-        max : Math.max
+        atan2 : { evaluate: Math.atan2, requireTypeMatch: true },
+        pow : { evaluate: Math.pow, requireTypeMatch: true },
+        min : { evaluate: Math.min, requireTypeMatch: false },
+        max : { evaluate: Math.max, requireTypeMatch: false }
```
Comment (lilleyse) [inline Source/Scene/Expression.js]: While this is more automated, I think its slightly overbuilt considering there are only 4 function. It should be fine to check the call name in `getEvaluateBinaryFunction` and apply the special cases.

**Comment by lilleyse** [comment]
> I'm not sure if it's my throw new DeveloperError notation that's wrong, but I don't see why all of my Error-throwing tests are failing :(

For those cases the error is thrown when the expression is evaluated, not when it's created. Take a look at https://github.com/AnalyticalGraphicsInc/cesium/blob/3d-tiles/Specs/Scene/ExpressionSpec.js#L815-L820 and others.

**Comment by lilleyse** [comment]
Also what errors are your receiving when running locally? Are you able to access the page at http://localhost:8080?

**Comment by lilleyse** [comment]
@Jane-Of-Art any updates here?

**Comment by Jane-Of-Art** [comment]
@lilleyse Apologies, we were on spring break last week! All rested now and will update this ticket later this week :)

**Comment by Jane-Of-Art** [comment]
@lilleyse 

Vector support for binary and ternary should all be here now! Your comments have been addressed and all tests should work. 

The tests now run locally as well, so thanks for the `npm run build` suggestion :)

**Comment by Jane-Of-Art** [comment]
@lilleyse 

As far as I can tell, the build errors are all rendering related, and the ExpressionSpec has passed. Can you tell if it's an error in the files that I edited?

Thanks! 
Jane

**Comment by lilleyse** [comment]
One of those errors should be fixed in https://github.com/AnalyticalGraphicsInc/cesium/pull/5100, but yeah they seem unrelated to this code.

I'll review soon.

File: Specs/Scene/ExpressionSpec.js [hunk 6]
```diff
@@ -2143,6 +2292,19 @@ defineSuite([
         }).toThrowDeveloperError();
     });
 
+    it('throws if min function takes mismatching types', function() {
```
Comment (lilleyse) [inline Specs/Scene/ExpressionSpec.js]: `min` -> `max`

File: Specs/Scene/ExpressionSpec.js [hunk 7]
```diff
@@ -2143,6 +2292,19 @@ defineSuite([
...
+            expression.evaluate(frameState, undefined);
+        }).toThrowDeveloperError();
+
+        expression = new Expression('max(vec2(0,1),vec3(0,1,2))');
+        expect(function() {
+            expression.evaluate(frameState, undefined);
+        }).toThrowDeveloperError();
+    });
+
+
```
Comment (lilleyse) [inline Specs/Scene/ExpressionSpec.js]: Remove empty line.

File: Source/Scene/Expression.js [hunk 8]
```diff
@@ -792,14 +792,72 @@ define([
...
+            //>>includeEnd('debug');
+            return evaluate(left, right); // jshint ignore:line
         };
     }
 
     function getEvaluateTernaryFunction(call) {
         var evaluate = ternaryFunctions[call];
         return function(feature) {
-            return evaluate(this._left.evaluate(feature), this._right.evaluate(feature), this._test.evaluate(feature));
+
```
Comment (lilleyse) [inline Source/Scene/Expression.js]: Remove empty line.

**Comment by Jane-Of-Art** [comment]
@lilleyse @Dylan-Brown 

Comments fixed, functionality for length and normalize added. I can delete the length functionality this week if you prefer the other implementation, but Dylan wanted me to implement it this week and I was pretty much done so I figured I'd include it for now.

Also, I noticed that there were shaderExpression tests at the bottom of the expressionSpec. Is that something we should worry about writing?

Thanks!
Jane

File: Specs/Scene/ExpressionSpec.js (hunk 6 above)
Comment (Jane-Of-Art) [inline Specs/Scene/ExpressionSpec.js]: Changed min -> max in the correct location below. I believe it should be min here

File: Specs/Scene/ExpressionSpec.js [hunk 9]
```diff
@@ -2011,11 +2011,61 @@ defineSuite([
 
     it('throws if fract function takes an invalid number of arguments', function() {
         expect(function() {
-            return new Expression('log2()');
+            return new Expression('fract()');
         }).toThrowDeveloperError();
```
Comment (Jane-Of-Art) [inline Specs/Scene/ExpressionSpec.js]: Fixed what I believe was a copy & paste error here

File: Source/Scene/Expression.js [hunk 10]
```diff
@@ -124,6 +126,42 @@ define([
...
+        }
+    }
+
+    function normalize(arg) {
+        var length;
+        if (typeof arg === 'number') {
+            return 1.0;
+        } else if (arg instanceof Cartesian2) {
+            length = Cartesian2.magnitude(arg);
+            return Cartesian2.fromElements(arg.x / length, arg.y / length, ScratchStorage.getCartesian2());
```
Comment (lilleyse) [inline Source/Scene/Expression.js]: Use `Cartesian2.normalize` instead.

File: Specs/Scene/ExpressionSpec.js (hunk 9 above)
Comment (lilleyse) [inline Specs/Scene/ExpressionSpec.js]: Ah yes, thanks for fixing that!

Comment (lilleyse) [inline Specs/Scene/ExpressionSpec.js]: Ah... looks like this was already fixed in `3d-tiles` actually.

**Comment by lilleyse** [comment]
> I can delete the length functionality

It's fine to take yours since you already finished it, I didn't know the other PR would come so soon. I like how `length` is separated out, and it makes me think we could do that for a lot of the special case functions, not just here but in #5040 .

> Also, I noticed that there were shaderExpression tests at the bottom of the expressionSpec. Is that something we should worry about writing?

Yeah I forgot about that, please add tests there as well.

@glee2244 these comments also apply to #5040

**Comment by Jane-Of-Art** [comment]
@lilleyse 

Everything should be good now! Should I resolve the conflicts? These mostly arise from the fact that length is in there twice.

**Comment by lilleyse** [comment]
Yes resolve the conflicts, just keep your versions.

**Comment by lilleyse** [comment]
@Jane-Of-Art Still ready to go, just fix the merge conflicts.

**Comment by Jane-Of-Art** [comment]
@lilleyse Rebased code!

**Comment by lilleyse** [comment]
I think something went wrong, the PR now has a lot of extraneous files. Instead of a rebase could you try a merge? Or cherry-pick your commits on top of origin/3d-tiles.

**Comment by lilleyse** [comment]
Also sorry to add more work here, but https://github.com/AnalyticalGraphicsInc/cesium/pull/5040 is now merged and introduced some conflicts.

**Comment by pjcozzi** [comment]
@Jane-Of-Art @lilleyse what is the plan here?  Do we want to get this into `3d-tiles` soon so we get that branch in shape for master?

**Comment by lilleyse** [comment]
@Jane-Of-Art hopefully you don't mind but I fixed up the PR and fixed the merge conflicts. As before the code looks good, so this is ready to merge.

**Pull request merged**